pkg/zip.go

```diff
@@ -56,68 +57,72 @@ func ArchiveWithFilterFunc(inFilePath string, filter func(string) bool) (result
 			return
 		}
 
-		var relativeFilePath string
-		if relativeFilePath, err = filepath.Rel(basePath, filePath); err != nil {
-			logging.Get().Warnf("Relative Filepath Error: %v", err)
+		// we only want the immediate directory and the filename for the archive path
+		var archivePath string
+		if archivePath, err = filepath.Rel(inFilePath, filePath); err != nil {
+			logging.Get().Warnf("ArchivePath Error: %v", err)
 			return
 		}
 
-		encapsulatingDirectory, fileName := filepath.Split(filePath)
-		encapsulatingFolder := filepath.Base(encapsulatingDirectory)
-
-		// we only want the immediate directory and the filename for the archive path
-		// so we are going to truncate the archive path
-		archivePath := path.Join(encapsulatingFolder, fileName)
+		if strings.HasPrefix(archivePath, ".") {
+			logging.Get().Debugf(color.Gray.Sprintf("Ignoring hidden file: %v", filePath))
+			return
+		}
 
-		if strings.HasPrefix(archivePath, ".") || !filter(relativeFilePath) {
-			// todo: fix this; it's not properly filtering off of the low level directory and the filename
-			logging.Get().Debugf(color.Gray.Sprintf("Ignoring: %v", filePath))
+		if !filterKeep(archivePath) {
+			logging.Get().Debugf(color.Gray.Sprintf("Ignoring filtered file: %v", filePath))
 			return
 		}
 
-		// spin off a thread archiving the file
-		eg.Go(func() error {
+		// spin off a thread archiving each file
+		g.Go(func() (err error) {
 			logging.Get().Tracef("Processing: %v => %v", color.Green.Sprint(filePath), color.Yellow.Sprint(archivePath))
-			if bytes, err := ioutil.ReadFile(filePath); err != nil {
+			var fileBytes []byte
+			fileBytes, err = os.ReadFile(filePath)
+			if err != nil {
 				logging.Get().Warnf("Error Processing %v: %v", filePath, err)
-				return err
-			} else {
-				ch <- archiveSuccess{
-					Bytes:    bytes,
-					FilePath: archivePath,
-				}
+				return
+			}
+			success := archiveSuccess{
+				Bytes:    fileBytes,
+				FilePath: archivePath,
 			}
-			return nil
+			select {
```

Comment:
I think os.ReadFile(filePath) block for reading and then the go routine ends with the select, but I think ch <- success will happen all the time and it won't evaluate ctx.Done. So, I think that if there is no ctx.Done then We can add to the channel

select {
	case <-ctx.Done():
		return ctx.Err()
	default:
		ch <- success
	}

Reply:
The `ch <- success` needs to be in the `case` statement per the docs. I believe that differentiates the channel write from the context cancel events. https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-Pipeline